apollo-federation/src/connectors/json_selection/methods/common.rs

```diff
@@ -0,0 +1,87 @@
+use shape::Shape;
+
+#[allow(dead_code)]
+pub(crate) fn is_comparable_shape_combination(shape1: &Shape, shape2: &Shape) -> bool {
+    if Shape::float([]).accepts(shape1) {
+        Shape::float([]).accepts(shape2) || shape2.accepts(&Shape::unknown([]))
+    } else if Shape::string([]).accepts(shape1) {
+        Shape::string([]).accepts(shape2) || shape2.accepts(&Shape::unknown([]))
+    } else if shape1.accepts(&Shape::unknown([])) {
+        Shape::float([]).accepts(shape2)
+            || Shape::string([]).accepts(shape2)
+            || shape2.accepts(&Shape::unknown([]))
+    } else {
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[rstest::rstest]
+    #[case(Shape::float([]), Shape::float([]), true)]
+    #[case(Shape::float([]), Shape::unknown([]), true)]
+    #[case(Shape::float([]), Shape::name("test", []), true)]
+    #[case(Shape::float([]), Shape::int([]), true)]
+    #[case(Shape::string([]), Shape::string([]), true)]
+    #[case(Shape::string([]), Shape::unknown([]), true)]
+    #[case(Shape::string([]), Shape::name("test", []), true)]
+    #[case(Shape::unknown([]), Shape::float([]), true)]
+    #[case(Shape::unknown([]), Shape::int([]), true)]
+    #[case(Shape::unknown([]), Shape::string([]), true)]
+    #[case(Shape::unknown([]), Shape::unknown([]), true)]
+    #[case(Shape::unknown([]), Shape::name("test", []), true)]
+    #[case(Shape::name("test", []), Shape::float([]), true)]
+    #[case(Shape::name("test", []), Shape::string([]), true)]
+    #[case(Shape::name("test", []), Shape::unknown([]), true)]
+    #[case(Shape::name("test", []), Shape::int([]), true)]
+    #[case(Shape::name("test", []), Shape::name("test", []), true)]
+    #[case(Shape::int([]), Shape::float([]), true)]
+    #[case(Shape::int([]), Shape::int([]), true)]
+    #[case(Shape::int([]), Shape::name("test", []), true)]
+    #[case(Shape::int([]), Shape::unknown([]), true)]
+    fn test_is_comparable_shape_combination_2_positive_cases(
+        #[case] shape1: Shape,
+        #[case] shape2: Shape,
+        #[case] expected: bool,
+    ) {
+        let actual = is_comparable_shape_combination(&shape1, &shape2);
+        assert_eq!(expected, actual);
```

Comment:
```suggestion
        assert!(is_comparable_shape_combination(&shape1, &shape2));
        assert!(is_comparable_shape_combination(&shape2, &shape1));
```
The opposite should always be true, yeah? Might mean we can eliminate some `rstest` up above, too, so it just lists the two shapes which should be comparable.